optional-chaining

aminya opened the pull request:
an attempt to fix #1909 
- Properties optional chaining support
- Fixes cli tests

Note: both properties and function call optional chaining are part of the proposal:
https://babeljs.io/docs/en/babel-plugin-proposal-optional-chaining

But this PR only implements properties optional chaining.

eventualbuddha commented:
Thanks! This looks like a good start, but needs tests. Some examples:

- `a?.b`
- `a?.[b]`
- `a?.b?.c`
- `a?.b()`
- `a?.()`
- `a.b?.()`
- `a?.b?.()`

aminya commented:
I don't know how you can pass compiler options to the tests! @eventualbuddha Could you help with that?

 This PR just skips properties conversion, so the tests are trivial.

aminya commented:
@eventualbuddha I am not familiar with the structure of the tests. I have enabled edit access. Could you help with setting up the test?

eventualbuddha commented:
@aminya I added some commits that should at least point you in the right direction. You’ll need to finish adding the expected soak test case results when using optional chaining in addition to the actual implementation. Let me know if you have more questions.

aminya commented:
@eventualbuddha I fixed the CLI tests that were failing for me.

I added the optional chaining test starting from https://github.com/decaffeinate/decaffeinate/pull/1969/commits/e747cf071ad92435d220702edc240408abbe6ba6.

Some of them fail because `AddVariableDeclarationsStage` fails to parse that. Probably you should add https://babeljs.io/docs/en/next/babel-plugin-syntax-optional-chaining.html to your parser.

krissrex commented:
Do you have any tests that try to convert coffee `someArray?[0]` to js `someArray?.[0]` ?

Ref https://github.com/decaffeinate/decaffeinate/issues/1909#issuecomment-656015308 and https://github.com/decaffeinate/decaffeinate/issues/1909#issuecomment-656033291

eventualbuddha commented:
This turned out to be not that straightforward. You can't simply leave the CoffeeScript code alone since in some cases the JavaScript equivalent has an extra `.`. Still, that part was pretty easy. The harder part is that CoffeeScript soaked expressions can be used in place that JavaScript optional chaining cannot, such as in an assignee: `a?.b = 1` is valid CoffeeScript, but not JavaScript. To address this I added `isInsideAssignee` to check whether a node should be considered part of an assignee or not, and to bail early if it is. Otherwise you'd end up with a parse error in a later stage.

As for actually landing this, I'd like to make this PR focus only on optional chaining. I removed some of the unrelated commits already, but it's still failing for me locally. Can you remove the other changes you made, @aminya? They'd be welcome as separate PRs.

eventualbuddha commented on src/stages/main/patchers/SoakedFunctionApplicationPatcher.ts:
```diff
@@ -151,6 +158,13 @@ export default class SoakedFunctionApplicationPatcher extends FunctionApplicatio
+  // patchDynamicMethodCallOptionalChaining(fn: MemberAccessOpPatcher): void {
+  //   // TODO
+  // }
+
```
Remove this.

eventualbuddha commented on src/stages/main/patchers/SoakedFunctionApplicationPatcher.ts:
```diff
@@ -121,6 +121,13 @@ export default class SoakedFunctionApplicationPatcher extends FunctionApplicatio
+  // patchMethodCallOptionalChaining(fn: MemberAccessOpPatcher): void {
+  //   // TODO
+  // }
+
```
Remove this.

eventualbuddha commented on src/stages/main/patchers/SoakedMemberAccessOpPatcher.ts:
```diff
@@ -9,7 +9,7 @@ const GUARD_HELPER = `function __guard__(value, transform) {
 
 export default class SoakedMemberAccessOpPatcher extends MemberAccessOpPatcher {
-  _shouldSkipSoakPatch = false;
+  _shouldSkipSoakPatch = this.options.optionalChaining;
```
Revert this.

eventualbuddha commented on test/cli_test.ts:
```diff
@@ -4,19 +4,24 @@ import { existsSync, readFileSync } from 'fs';
 import { copySync } from 'fs-extra';
 
 import stripSharedIndent from '../src/utils/stripSharedIndent';
+import { toUnix } from './utils/toUnix';
```
These changes belong in another PR.

eventualbuddha commented:
:tada: This PR is included in version 6.1.0 :tada:

The release is available on:
- [npm package (@latest dist-tag)](https://www.npmjs.com/package/decaffeinate/v/6.1.0)
- [GitHub release](https://github.com/decaffeinate/decaffeinate/releases/tag/v6.1.0)

Your **[semantic-release](https://github.com/semantic-release/semantic-release)** bot :package::rocket:

aminya commented:
Thanks for the update. I am going to make another pull request that fixed the CLI tests. I am not sure what you have removed at this point. Nothing is apparent from the commits. Fortunately, I have a local backup.

eventualbuddha commented:
The commits are still here on this PR, or retrieving them locally works too. Thanks for getting this started!
